### DISCUSSION TITLE
SQLAlchemy - PostgreSQL : Why name in made required in Index Construct

### DISCUSSION THREAD

**Discussion by simonnelse** [sqlalchemy/sqlalchemy#7469] [Usage Questions] [answered]
Hi, 

To create indexes which encompass more than one column we need to use the [`Index`](https://docs.sqlalchemy.org/en/14/core/constraints.html#sqlalchemy.schema.Index)  construct in SQLAlchemy. The developer **must mandatorily** provide a name for the index in SQLAlchemy.  But PostgreSQL supports creating indexes [without providing a name](https://www.postgresql.org/docs/14/sql-createindex.html), ie If the name is omitted, PostgreSQL chooses a suitable name based on the parent table's name and the indexed column name(s).

```sql
-- postgres creates an index by the name `tableName_col1_col2_idx` 
CREATE INDEX ON tableName(col1, col2) 
```

So, Is there a reason for making name required in SQLAlchemy `Index` construct.

**Comment by CaselIT** [comment]
Hi,

Postgresql is just one of the db sqlalchemy supports. Have you tried passing none as the name of the index to see it it's omitted from the SQL when using postgresql?

**Reply by simonnelse** [reply]
Hello @CaselIT, 

```python
# SQLAlchemy creates an index by the name ix_my_table_id
Index(None, my_table.c.id)

# again tries to create an index by the name ix_my_table_id, but throws 'relation already exists' error
Index(None, my_table.c.id, postgresql_where = my_table.c.value > 10)
```

PostgreSQL handles this scenario by suffixing a sequential number (like `ix_my_table_id`, `ix_my_table_id1`, `ix_my_table_id2`, . . .) at the end of index name.

Checked this scenario with MySQL _(one of the DB supported by SQLAlchemy)_, index name is mandatory in MySQL.

**Reply by CaselIT** [reply]
I think that if name is none sqlalchemy tries to use the naming convention to generate one.

I don't know if the naming convention can be suppressed, you would have to look in the docs

**Reply by zzzeek** [reply]
the proposal is to supply a constant NONE_NAME, see https://github.com/sqlalchemy/sqlalchemy/issues/4289#issuecomment-441942080 

so this is a fairly straightforward request, would need more docs etc.

**Comment by zzzeek** [comment, marked as answer]
> So, Is there a reason for making name required in SQLAlchemy `Index` construct.

Most supported databases other than PostgreSQL require a name for indexes up front, and SQLAlchemy does not want to be in the business of generating names, so this drove the API design of the `Index` construct.    However, the fact that PG indexes can be created without a name will be supported if/when issue #4289 is implemented, so this has been accepted as a use case request.      

as a workaround, please use the [DDL](https://docs.sqlalchemy.org/en/14/core/ddl.html?highlight=ddl#sqlalchemy.schema.DDL) construct to emit this PostgreSQL-specific syntax on table create for now.
